sh: enable lexer styles for command substitution and related styling

techee opened the pull request:
Before the patch:

<img width="392" alt="Screenshot 2024-11-26 at 22 23 52" src="https://github.com/user-attachments/assets/d8a858cf-298f-4008-89ba-731c3e0a7de1">

After the patch:

<img width="392" alt="Screenshot 2024-11-26 at 22 23 17" src="https://github.com/user-attachments/assets/1b069e87-1365-40bc-bdd4-a15eaf01e340">

Fixes #1754
Fixes #1821

b4n commented:
Although it seems like a pretty obvious thing to want, we need to be a bit careful I'm pretty sure I remember those had pretty serious limitations that made things worse in real-world scripts. It might have been fixed which would be awesome, but we should check fairly comprehensively.
I'll try to throw most of my random scripts a it to see if I find any issue.

techee commented:
> I'll try to throw most of my random scripts a it to see if I find any issue.

I tried scripts in the Geany's `scripts` directory yesterday and at least those seemed alright.

Also, if there are problems with `lexer.bash.command.substitution`, it also has possible value `2`. From the description it's not completely clear what the difference is, I used `1` because it colorized comments and strings inside `$()` unlike `2`, but if there are problems, we could use it as well.

zufuliu commented:
`lexer.bash.command.substitution=2` (for C++ preprocessor inactive block like tracking) was added in https://github.com/ScintillaOrg/lexilla/issues/153

eht16 commented:
To me it looks fine, I didn't find a script yet where the highlighting is obviously broken but will keep an eye on it during day by day usage.

If we cannot solve all doubts, I think it would still be a good change to include because disabling the settings if they are present might be easier for users than getting aware that the settings can be added.

techee commented:
> We should probably set lexer.bash.styling.inside.heredoc=1 as well

OK, done.

> I'm not sure I really like lexer.bash.styling.inside.string=1. Yes, it's great when there are complex expansions. But I find it lowers usability when quoting single variables, which is overly common, as it makes the fact it's quoted a lot harder to notice.

I have no problem with having it set to 0 by default - I just added this for completeness, the main part was `lexer.bash.command.substitution=1` to fix the two bug reports. I've added a commit (to be squashed) setting ` lexer.bash.styling.inside.string=0` (I kept it in the config file so it's easier to discover this settings for users).

techee commented:
Merged squashed version.
